liboptv/src/multimed.c

```diff
@@ -0,0 +1,656 @@
+ 
+  Zmin -= fmod (Zmin, rw);
+  Zmax += (rw - fmod (Zmax, rw));
+  Zmin_t=Zmin;
+  Zmax_t=Zmax;
+
+  /* intersect with image vertices rays */
+
+  for (i = 0; i < 2; i ++) {
+     for (j = 0; j < 2; j++) {
+  
+      pixel_to_metric (&x, &y, xc[i], yc[j], cpar);
+  
+
+      x = x - cal->int_par.xh;
+      y = y - cal->int_par.yh;
+  
+      correct_brown_affin (x, y, cal->added_par, &x,&y);
+            
+
+      /* ray_tracing(x,y, Ex, I, G, mmp, &X1, &Y1, &Z1, &a, &b, &c); */
+      ray_tracing(x,y, cal, *(cpar->mm), pos, a);
+  
+             
+  
+      /* Z = Zmin;   X = X1 + (Z-Z1) * a/c;   Y = Y1 + (Z-Z1) * b/c; */
+      Z = Zmin;   
+      X = pos[0] + (Z - pos[2]) * a[0]/a[2];   
+      Y = pos[1] + (Z - pos[2]) * a[1]/a[2];
+            
+      xyz[0] = X; xyz[1] = Y; xyz[2] = Z;
+  
+      /* trans_Cam_Point(Ex,mmp,G,X,Y,Z,&Ex_t,&X_t,&Y_t,&Z_t,&cross_p,&cross_c); */
+      trans_Cam_Point(cal->ext_par, *(cpar->mm), cal->glass_par, xyz, \
+      &Ex_t, xyz_t, (double *)cross_p, (double *)cross_c);
+
+      X_t = xyz_t[0]; Y_t = xyz_t[1]; Z_t = xyz_t[2];      
+        
+      if( Z_t < Zmin_t ) Zmin_t = Z_t;
+      if( Z_t > Zmax_t ) Zmax_t = Z_t;
+        
+
+      R = sqrt (( X_t - Ex_t.x0 ) * ( X_t - Ex_t.x0 )
+              + ( Y_t - Ex_t.y0 ) * ( Y_t - Ex_t.y0 )); 
+          
+
+
+      if (R > Rmax) Rmax = R;
+                  
+      /* Z = Zmax;   X = X1 + (Z-Z1) * a/c;   Y = Y1 + (Z-Z1) * b/c; */
+      Z = Zmax;   
+      X = pos[0] + (Z - pos[2]) * a[0]/a[2];   
+      Y = pos[1] + (Z - pos[2]) * a[1]/a[2];
+  
+      xyz[0] = X; xyz[1] = Y; xyz[2] = Z;
+
+      /* trans_Cam_Point(Ex,mmp,G,X,Y,Z,&Ex_t,&X_t,&Y_t,&Z_t,&cross_p,&cross_c); */
+      trans_Cam_Point(cal->ext_par, *(cpar->mm), cal->glass_par, xyz,\
+      &Ex_t, xyz_t, (double *)cross_p, (double *)cross_c);
+  
+      X_t = xyz_t[0]; Y_t = xyz_t[1]; Z_t = xyz_t[2];
+  
+
+      if( Z_t < Zmin_t ) Zmin_t = Z_t;
+      if( Z_t > Zmax_t ) Zmax_t = Z_t;
+
+
+      R = sqrt (( X_t - Ex_t.x0 ) * ( X_t - Ex_t.x0 )
+              + ( Y_t - Ex_t.y0 ) * ( Y_t - Ex_t.y0 )); 
+
+
+      if (R > Rmax) Rmax = R;
+  
+    }
+  }
+
+
+
+  /* round values (-> enlarge) */
+  Rmax += (rw - fmod (Rmax, rw));
+
+
+  /* get # of rasterlines in r,z */
+  // BUG: fixed by Ad Holten 
+  // nr = Rmax/rw + 1;
+  // nz = (Zmax_t - Zmin_t)/rw + 1;
+
+
+  /* get # of rasterlines in r,z */
+  nr = (int)(Rmax/rw + 1);
+  nz = (int)((Zmax_t-Zmin_t)/rw + 1);
+
+
+  /* create two dimensional mmlut structure */
+  xyz[0] = X; xyz[1] = Y; xyz[2] = Z;
+
+  trans_Cam_Point(cal->ext_par, *(cpar->mm), cal->glass_par, xyz,\
+      &Ex_t, xyz_t, (double *)cross_p, (double *)cross_c);
+  
+  X_t = xyz_t[0]; Y_t = xyz_t[1]; Z_t = xyz_t[2];
+
+  cal->mmlut.origin.x = Ex_t.x0;
+  cal->mmlut.origin.y = Ex_t.y0;
+  cal->mmlut.origin.z = Zmin_t;
+  cal->mmlut.nr = nr;
+  cal->mmlut.nz = nz;
+  cal->mmlut.rw = rw;
+  // if (cal.mmlut.data != NULL)			// preventing memory leaks, ad holten, 04-2013

```

Comment:
This seems important actually. mmlut.data will not necessarily be NULL though, because we don't initialize it anywhere. Maybe the structure should be handled somehow in read_calibration.

Max is working on some related things, I'll check if he has a good base to work from.